src/array/cpu/spmm.h

```diff
@@ -0,0 +1,347 @@
+  IdType* argW = Op::use_rhs? static_cast<IdType*>(arge->data) : nullptr;
+#pragma omp parallel for
+  for (IdType rid = 0; rid < csr.num_rows; ++rid) {
+    const IdType row_start = indptr[rid], row_end = indptr[rid + 1];
+    DType* out_off = O + rid * dim;
+    IdType* argx_off = argX + rid * dim;
+    IdType* argw_off = argW + rid * dim;
+    for (int64_t k = 0; k < dim; ++k) {
+      DType accum = Cmp::zero;
+      IdType ax = 0, aw = 0;
+      for (IdType j = row_start; j < row_end; ++j) {
+        const IdType cid = indices[j];
+        const IdType eid = has_idx? edges[j] : j;
+        const int64_t lhs_add = bcast.use_bcast ? bcast.lhs_offset[k] : k;
+        const int64_t rhs_add = bcast.use_bcast ? bcast.rhs_offset[k] : k;
+        const DType* lhs_off = Op::use_lhs? X + cid * lhs_dim + lhs_add : nullptr;
+        const DType* rhs_off = Op::use_rhs? W + eid * rhs_dim + rhs_add : nullptr;
+        const DType val = Op::Call(lhs_off, rhs_off);
+        if (Cmp::Call(accum, val)) {
+          accum = val;
+          if (Op::use_lhs)
+            ax = cid;
+          if (Op::use_rhs)
+            aw = eid;
+        }
+      }
+      out_off[k] = accum;
+      if (Op::use_lhs)
+        argx_off[k] = ax;
+      if (Op::use_rhs)
+        argw_off[k] = aw;
+    }
+  }
+}
+
+/*!
+ * \brief CPU kernel of SpMM-Min/Max on Coo format.
+ * \param bcast Broadcast information.
+ * \param coo The Coo matrix.
+ * \param ufeat The feature on source nodes.
+ * \param efeat The feature on edges.
+ * \param out The result feature on destination nodes.
+ * \param argu Arg-Min/Max on source nodes.
+ * \param arge Arg-Min/Max on edges.
+ * \note it uses node parallel strategy, different threads are responsible
+ *       for the computation of different nodes. To avoid possible data hazard,
+ *       we use atomic operators in the reduction phase.
+ */
+template <typename IdType, typename DType, typename Op, typename Cmp>
+void SpMMCmpCoo(
+    const BcastOff& bcast,
+    const aten::COOMatrix& coo,
+    NDArray ufeat, NDArray efeat,
+    NDArray out, NDArray argu, NDArray arge) {
+  const bool has_idx = !aten::IsNullArray(coo.data);
+  const IdType* row = static_cast<IdType*>(coo.row->data);
+  const IdType* col = static_cast<IdType*>(coo.col->data);
+  const IdType* edges = has_idx? static_cast<IdType*>(coo.data->data) : nullptr;
+  const DType* X = Op::use_lhs? static_cast<DType*>(ufeat->data) : nullptr;
+  const DType* W = Op::use_rhs? static_cast<DType*>(efeat->data) : nullptr;
+  const int64_t dim = bcast.out_len,
+                lhs_dim = bcast.lhs_len,
+                rhs_dim = bcast.rhs_len;
+  DType* O = static_cast<DType*>(out->data);
+  IdType* argX = Op::use_lhs? static_cast<IdType*>(argu->data) : nullptr;
+  IdType* argW = Op::use_rhs? static_cast<IdType*>(arge->data) : nullptr;
+  const int64_t nnz = coo.row->shape[0];
+  // fill zero elements
+#pragma omp parallel for
+  for (IdType i = 0; i < out.NumElements(); ++i) {
+    O[i] = Cmp::zero;
```

Comment:
use memset?

Reply:
We don't know the data type of O, maybe std::fill is a better solution.
btw, I agree that it's not necessary to use openmp here.